source/util/FileUtils.cpp

```diff
@@ -296,6 +303,35 @@ bool FileUtils::CreateDirectoryWithPermissions(const char *dirPath, mode_t permi
     return false;
 }
 
+bool FileUtils::DirectoryExists(const std::string &dirPath)
+{
+    struct stat dirInfo;
+    if (stat(dirPath.c_str(), &dirInfo) != 0)
+    {
+        // Ignore permission errors.
+        return false;
+    }
+    return S_ISDIR(dirInfo.st_mode);
+}
+
+bool FileUtils::CreateEmptyFileWithPermissions(const string &filename, mode_t permissions)
+{
+    auto fd = open(filename.c_str(), O_CREAT | O_EXCL, permissions);
```

Comment:
I think we should extract the expanded absolute path of the file before calling `open` to create one.

Reply:
Done.